Sources/Noora/Components/ProgressBarStep.swift

```diff
@@ -0,0 +1,149 @@
+    let errorMessage: String?
+    let task: (@escaping (Double) -> Void) async throws -> V
+    let theme: Theme
+    let terminal: Terminaling
+    let renderer: Rendering
+    let standardPipelines: StandardPipelines
+    let spinner: Spinning
+    let logger: Logger?
+
+    private let complete = "█"
+    private let incomplete = "▒"
+    private let width = 30
+
+    init(
+        message: String,
+        successMessage: String?,
+        errorMessage: String?,
+        task: @escaping (@escaping (Double) -> Void) async throws -> V,
+        theme: Theme,
+        terminal: Terminaling,
+        renderer: Rendering,
+        standardPipelines: StandardPipelines,
+        spinner: Spinning = Spinner(),
+        logger: Logger?
+    ) {
+        self.message = message
+        self.successMessage = successMessage
+        self.errorMessage = errorMessage
+        self.task = task
+        self.theme = theme
+        self.terminal = terminal
+        self.renderer = renderer
+        self.standardPipelines = standardPipelines
+        self.spinner = spinner
+        self.logger = logger
+    }
+
+    func run() async throws -> V {
+        if terminal.isInteractive {
+            return try await runInteractive()
+        } else {
+            return try await runNonInteractive()
+        }
+    }
+
+    func runInteractive() async throws -> V {
+        let start = DispatchTime.now()
+
+        defer {
+            spinner.stop()
+        }
+
+        var spinnerIcon: String?
+        var lastProgress = 0.0
+
+        spinner.spin { icon in
+            spinnerIcon = icon
+            render(progress: lastProgress, icon: spinnerIcon ?? "ℹ︎")
+        }
+
+        do {
+            let result = try await task { progress in
+                lastProgress = progress
+            }
+            renderer.render(
+                .progressCompletionMessage(
+                    (successMessage ?? message).hexIfColoredTerminal(theme.primary, terminal).boldIfColoredTerminal(terminal),
+                    timeString: timeString(start: start),
+                    theme: theme,
+                    terminal: terminal
+                ),
+                standardPipeline: standardPipelines.output
+            )
+            logger?.debug("'\(message)' succeeded with '\(successMessage ?? message)'")
+            return result
+        } catch {
+            renderer.render(
+                .progressErrorMessage(
+                    (errorMessage ?? message).hexIfColoredTerminal(theme.danger, terminal).boldIfColoredTerminal(terminal),
+                    timeString: timeString(start: start),
+                    theme: theme,
+                    terminal: terminal
+                ),
+                standardPipeline: standardPipelines.error
+            )
+            logger?.error("'\(message)' failed with '\(errorMessage ?? message)'")
+            throw error
+        }
+    }
+
+    func runNonInteractive() async throws -> V {
```

Comment:
The non-interactive version doesn't convey progress:
![non-interactive](https://github.com/user-attachments/assets/92eaa611-0fee-423f-9502-83809b75bcc0)

What if we use multiple lines for that? I believe that's what CLIs generally do:

```bash
ℹ︎ Loading manifests
ℹ︎ Loading manifests ██████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒ 20%
ℹ︎ Loading manifests █████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒ 30%
ℹ︎ Loading manifests ██████████████████████▒▒▒▒▒▒▒▒ 60%
✔︎ Manifests loaded [4.3s]
```

Alternatively, we get rid of the bar and just show the numeric value:

```bash
ℹ︎ Loading manifests
ℹ︎ Loading manifests 20%
ℹ︎ Loading manifests 30%
ℹ︎ Loading manifests 60%
✔︎ Manifests loaded [4.3s]
```

Reply:
good idea, done 👍